crates/flox-rust-sdk/src/models/environment.rs

```diff
@@ -151,21 +152,40 @@ impl<S: TransactionState> PathEnvironment<S> {
     }
 
     /// Replace the contents of this environment's `.flox` with that of another environment's `.flox`
+    ///
+    /// This may copy build symlinks, so the assumption is that building self
+    /// will result in the same out link as building the replacement.
     pub fn replace_with(
         &mut self,
         replacement: PathEnvironment<Temporary>,
     ) -> Result<(), EnvironmentError2> {
-        fs_extra::dir::move_dir(
-            replacement.path,
-            &self.path,
-            &fs_extra::dir::CopyOptions::new()
-                .overwrite(true)
-                .content_only(true),
-        )
-        .expect("replace origin");
-        // std::fs::create_dir(&replacement.path).expect("recreate temp dir");
+        let transaction_backup = self
+            .path
+            .parent()
+            .expect("path shouldn't be /")
+            .join(format!("{}.tmp", self.name().0));
+        if transaction_backup.exists() {
+            return Err(EnvironmentError2::PriorTransaction(transaction_backup));
+        }
+        fs::rename(&self.path, &transaction_backup)
+            .map_err(EnvironmentError2::BackupTransaction)?;
+        fs::rename(replacement.path, &self.path).map_err(EnvironmentError2::Move)?;
+        fs::remove_dir_all(transaction_backup).map_err(EnvironmentError2::RemoveBackup)?;
         Ok(())
     }
+
+    /// Where to link a built environment to. The parent directory may not exist.
+    ///
+    /// When used as a lookup signals whether the environment has *at some point* been built before
+    /// and is "activatable". Note that the environment may have been modified since it was last built.
+    ///
+    /// Mind that an existing out link does not necessarily imply that the environment
+    /// can in fact be built.
+    pub fn out_link(&self, system: impl AsRef<str> + Send) -> PathBuf {
```

Comment:
If this method has all these caveats, maybe this method should be private instead so that it can't be accidentally misused by a consumer.